### PR TITLE
Added bxt_viewmodel_fov

### PR DESCRIPTION
**Pull request by chinese-soup**
Based on the previous OpenAG implementation, therefore same values apply, to reiterate:
- `default_fov 90` && `bxt_viewmodel_fov 0` ==> no application, FOV of weapon same as default_fov's value
- `default_fov 90` && `bxt_viewmodel_fov 90` ==> yes, apply frustum, but FOV of weapon is the same as on default_fov 90
- `default_fov 90` && `bxt_viewmodel_fov 130` ==> yes, apply frustum, FOV of weapon is the same as it would be with default_fov 130
etc.

**Note:**
```c++
// If the current's frame FOV is not default_fov, we are zoomed in, in that case don't override frustum
if (desired_viewmodel_fov > 0 && desired_viewmodel_fov < 179 && currentRenderFOV == CVars::default_fov.GetFloat())
```

If `currentRenderFOV` is never updated because of the `HUD_UpdateClientData` function not being hooked, the `bxt_viewmodel_fov` CVar will appear but its value is never gonna apply and `ORIG_R_DrawViewModel` is just gonna get called. 
I dunno how to make it so that the Cvar only registers when both `HUD_UpdateClientData` and `R_DrawViewModel` are available since they are from two different classes. :/

**Note 2:**
Missing Windows patterns, will take a look later unless somebody else wants to. ;)

**Note 3:**
Should this be considered a cheat (bxt_ch_viewmodel_fov)? Screenshots can be found here: https://github.com/YaLTeR/OpenAG/pull/107

### CODE REVIEW

**Comment by chinese-soup** [comment]
Drafting since I forgot about the attachments. :)

**Comment by hobokenn** [comment]
PATTERNS(R_DrawViewModel,
			"HL-SteamPipe",
			"55 8B EC 83 EC 50 D9 05 ?? ?? ?? ?? D8 1D ?? ?? ?? ?? 56 57 33 FF C7 45",
			"HL-4554",
			"83 EC ?? D9 05 ?? ?? ?? ?? D8 1D ?? ?? ?? ?? 56 57 33 FF C7 44"
		);
		
> If currentRenderFOV is never updated because of the HUD_UpdateClientData function not being hooked

It shouldn't be a problem because this function is always hooked, because it is exported.

File: BunnymodXT/modules/HwDLL.hpp [hunk 1]
```diff
@@ -42,6 +42,7 @@ class HwDLL : public IHookableNameFilterOrdered
 	HOOK_DECL(void, __cdecl, Cmd_Exec_f)
 	HOOK_DECL(void, __cdecl, R_DrawSequentialPoly, msurface_t *surf, int face)
 	HOOK_DECL(void, __cdecl, R_Clear)
+	HOOK_DECL(void, __cdecl, R_DrawViewModel)
```
Comment (hobokenn) [inline BunnymodXT/modules/HwDLL.hpp:45]: This should probably be int

File: BunnymodXT/modules/HwDLL.cpp [hunk 2]
```diff
@@ -4498,6 +4518,34 @@ HOOK_DEF_2(HwDLL, void, __cdecl, R_DrawSequentialPoly, msurface_t *, surf, int,
 	glDisable(GL_BLEND);
 }
 
+
```
Comment (YaLTeR) [inline BunnymodXT/modules/HwDLL.cpp:4521]: ah yes it's one of our favorites again

**Comment by YaLTeR** [comment]
Needs a conflict fix too

**Comment by chinese-soup** [comment]
> Needs a conflict fix too

Yeah ik, it's still a draft, cuz I need to apply all (maybe except for thirdperson shell ejection since this is meant for spedrunning single player games) the fixes tmp64 made to this code.

**Comment by chinese-soup** [comment]
May have fucked up the merge commit while resolving the changes a bit, but hopefully this should be alright now? I can squash now if it's needed.

File: BunnymodXT/modules/ClientDLL.cpp [hunk 3]
```diff
@@ -797,6 +797,8 @@ HOOK_DEF_2(ClientDLL, int, __cdecl, HUD_UpdateClientData, client_data_t*, pcldat
 	const auto norefresh = CVars::_bxt_norefresh.GetBool();
 	int (*ORIG_GetScreenInfo)(SCREENINFO *pscrinfo) = nullptr;
 
+	HwDLL::GetInstance().currentRenderFOV = pcldata->fov;
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.cpp:937]: Shouldn't this be after `ORIG_HUD_UpdateClientData` in case the client updates the FOV?

File: BunnymodXT/modules/HwDLL.cpp [hunk 4]
```diff
@@ -759,7 +759,7 @@ void HwDLL::FindStuff()
 		DEF_FUTURE(VGuiWrap2_ConPrintf)
 		DEF_FUTURE(R_DrawSequentialPoly)
 		DEF_FUTURE(R_Clear)
-		//DEF_FUTURE(R_DrawViewModel)
+		DEF_FUTURE(R_DrawViewModel)
```
Comment (YaLTeR) [inline BunnymodXT/modules/HwDLL.cpp:885]: Could you squash this commit into previous?

File: BunnymodXT/modules/ClientDLL.cpp [hunk 5]
```diff
@@ -63,6 +63,22 @@ extern "C" int __cdecl HUD_UpdateClientData(client_data_t* pcldata, float flTime
 {
 	return ClientDLL::HOOKED_HUD_UpdateClientData(pcldata, flTime);
 }
+
+extern "C" void __cdecl _ZN20CStudioModelRenderer21StudioCalcAttachmentsEv(void *thisptr)
+{
+	return ClientDLL::HOOKED_StudioCalcAttachments(thisptr);
+}
+extern "C" void __cdecl _Z15VectorTransformPKfPA4_fPf(float *in1, float *in2, float *out)
+//extern "C" void __cdecl _Z15VectorTransformPKfPA4_fPf(const vec3_t in1, float* in2, vec3_t out)
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.cpp]: Leftover comment? Also missing newline

File: BunnymodXT/modules/ClientDLL.cpp [hunk 6]
```diff
@@ -261,6 +283,57 @@ void ClientDLL::FindStuff()
...
+		// TODO: fix up these pattern hooks to look like those in serverDLL
+		EngineDevWarning("[client dll] Could not find StudioCalcAttachments.\n");
+		auto fStudioCalcAttachments = FindAsync(
+				ORIG_StudioCalcAttachments,
+				patterns::client::StudioCalcAttachments);
+		auto pattern = fStudioCalcAttachments.get();
+
+		if (ORIG_StudioCalcAttachments)
+		{
+			EngineDevMsg("[client dll] [WINDOWS] Found StudioCalcAttachments at %p (using pattern %s).\n", ORIG_VectorTransform, pattern->name());
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.cpp]: been there done that fixed in bxt-rs
> `ORIG_VectorTransform`

File: BunnymodXT/modules/ClientDLL.cpp [hunk 7]
```diff
@@ -63,6 +63,22 @@ extern "C" int __cdecl HUD_UpdateClientData(client_data_t* pcldata, float flTime
...
+{
+	return ClientDLL::HOOKED_StudioCalcAttachments(thisptr);
+}
+extern "C" void __cdecl _Z15VectorTransformPKfPA4_fPf(float *in1, float *in2, float *out)
+//extern "C" void __cdecl _Z15VectorTransformPKfPA4_fPf(const vec3_t in1, float* in2, vec3_t out)
+{
+	return ClientDLL::HOOKED_VectorTransform(in1, in2, out);
+}
+
+extern "C" void __cdecl _Z22EV_GetDefaultShellInfoP12event_args_sPfS1_S1_S1_S1_S1_S1_fff(event_args_t *args, float *origin, float *velocity, float *ShellVelocity, float *ShellOrigin,
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.cpp:76]: Is it intended that this isn't hooked on windows?

File: BunnymodXT/modules/ClientDLL.cpp [hunk 8]
```diff
@@ -685,6 +790,12 @@ HOOK_DEF_1(ClientDLL, void, __cdecl, V_CalcRefdef, ref_params_t*, pparams)
 
 	last_vieworg = pparams->vieworg;
 	last_viewangles = pparams->viewangles;
+	last_viewforward = pparams->forward;
+	last_viewright = pparams->right;
+	last_viewup = pparams->up;
+
+	pEngfuncs->Con_Printf("ORIGIN = %f %f %f \n", last_vieworg[0], last_vieworg[1], last_vieworg[2]);
+	pEngfuncs->Con_Printf("forward = %f %f %f\n ", last_viewforward[0], last_viewforward[1], last_viewforward[2]);
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.cpp]: leftover debug prints

File: BunnymodXT/modules/ClientDLL.hpp [hunk 9]
```diff
@@ -22,6 +23,11 @@ class ClientDLL : public IHookableNameFilter
 	HOOK_DECL(void, __cdecl, HUD_DrawTransparentTriangles)
 	HOOK_DECL(int, __cdecl, HUD_Key_Event, int down, int keynum, const char* pszCurrentBinding)
 	HOOK_DECL(int, __cdecl, HUD_UpdateClientData, client_data_t* pcldata, float flTime)
+	HOOK_DECL(void, __cdecl, StudioCalcAttachments, void* thisptr)
+	HOOK_DECL(void, __cdecl, VectorTransform, float *in1, float *in2, float *out)
+	HOOK_DECL(void, __cdecl, EV_GetDefaultShellInfo, event_args_t *args, float *origin, float *velocity, float *ShellVelocity, float *ShellOrigin,
+	          float *forward, float *right, float *up, float forwardScale, float upScale,float rightScale)
+	//HOOK_DECL(void, __cdecl, VectorTransform, const vec3_t in1, float* in2, vec3_t out)
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.hpp]: leftover?

File: BunnymodXT/modules/HwDLL.cpp [hunk 10]
```diff
@@ -163,6 +163,18 @@ extern "C" void __cdecl SV_WriteEntitiesToClient(client_t* client, void* msg)
...
+/*extern "C" void __cdecl VectorTransform(const vec3_t in1, float* in2, vec3_t out)
+{
+	return HwDLL::HOOKED_VectorTransform(in1, in2, out);
+}*/
+
+//extern "C" void __cdecl V_SetRefParams(void *header)
+/*extern "C" int __cdecl studioapi_GetCurrentEntity()
+{
+	return HwDLL::HOOKED_studioapi_GetCurrentEntity();
+}*/
```
Comment (YaLTeR) [inline BunnymodXT/modules/HwDLL.cpp]: leftover?

File: BunnymodXT/modules/HwDLL.cpp [hunk 11]
```diff
@@ -279,7 +291,10 @@ void HwDLL::Hook(const std::wstring& moduleName, void* moduleHandle, void* modul
 			ORIG_R_DrawViewModel, HOOKED_R_DrawViewModel,
 			ORIG_Mod_LeafPVS, HOOKED_Mod_LeafPVS,
 			ORIG_SV_AddLinksToPM_, HOOKED_SV_AddLinksToPM_,
-			ORIG_SV_WriteEntitiesToClient, HOOKED_SV_WriteEntitiesToClient);
+			ORIG_SV_WriteEntitiesToClient, HOOKED_SV_WriteEntitiesToClient,
+			ORIG_VectorTransform, HOOKED_VectorTransform
+		//	ORIG_studioapi_GetCurrentEntity, HOOKED_studioapi_GetCurrentEntity
```
Comment (YaLTeR) [inline BunnymodXT/modules/HwDLL.cpp]: leftover?

File: BunnymodXT/modules/HwDLL.cpp [hunk 12]
```diff
@@ -4520,18 +4556,18 @@ HOOK_DEF_2(HwDLL, void, __cdecl, R_DrawSequentialPoly, msurface_t *, surf, int,
...
+
 		glMatrixMode (GL_PROJECTION);
 		glLoadIdentity();
 		GLfloat w, h;
 		GLfloat _near = 3.0f;
 		GLfloat _far = 4096.0f;
-		float ScreenWidth = CustomHud::GetScreenInfo().iWidth;
-		float ScreenHeight = CustomHud::GetScreenInfo().iHeight;
+		int ScreenWidth = CustomHud::GetScreenInfo().iWidth;
+		int ScreenHeight = CustomHud::GetScreenInfo().iHeight;
```
Comment (YaLTeR) [inline BunnymodXT/modules/HwDLL.cpp:4714]: Seems like a change that should be put into the original commit instead.

File: BunnymodXT/modules/HwDLL.cpp [hunk 13]
```diff
@@ -4520,18 +4556,18 @@ HOOK_DEF_2(HwDLL, void, __cdecl, R_DrawSequentialPoly, msurface_t *, surf, int,
 
 HOOK_DEF_0(HwDLL, void, __cdecl, R_DrawViewModel)
 {
-	auto desired_viewmodel_fov = CVars::bxt_viewmodel_fov.GetFloat();
-
 	// If the current's frame FOV is not default_fov, we are zoomed in, in that case don't override frustum
-	if (desired_viewmodel_fov > 0 && desired_viewmodel_fov < 179 && currentRenderFOV == CVars::default_fov.GetFloat())
+	if (NeedViewmodelAdjustments())
 	{
+		auto desired_viewmodel_fov = CVars::bxt_viewmodel_fov.GetFloat();
```
Comment (YaLTeR) [inline BunnymodXT/modules/HwDLL.cpp:4706]: Unused?

File: BunnymodXT/modules/HwDLL.cpp [hunk 14]
```diff
@@ -4587,3 +4623,17 @@ HOOK_DEF_2(HwDLL, void, __cdecl, SV_WriteEntitiesToClient, client_t*, client, vo
...
+}*/
+
+HOOK_DEF_3(HwDLL, void, __cdecl, VectorTransform, const vec3_t, in1, float*, in2, vec3_t, out)
+{
+	/*ORIG_Con_Printf("AHOJ JSEM V STUIDO SET HEADER\n");
+	int viewentity = ORIG_studioapi_GetCurrentEntity();
+	ORIG_Con_Printf("INTEGER %d\n", viewentity);*/
+	ORIG_Con_Printf("AHOJ VectorTransform\n");
+	return ORIG_VectorTransform(in1, in2, out);
+}
```
Comment (YaLTeR) [inline BunnymodXT/modules/HwDLL.cpp:4832]: Leftover stuff

File: BunnymodXT/patterns.hpp [hunk 15]
```diff
@@ -496,6 +496,19 @@ namespace patterns
 		);
 	}
 
+	namespace client
+	{
+		PATTERNS(VectorTransform,
+			"HL-4554",
+			"8B 44 24 ?? 8B 4C 24 ?? 8B 54 24 ?? D9 40 ??"
```
Comment (YaLTeR) [inline BunnymodXT/patterns.hpp]: Don't leave ?? in the end, it's no use (bxt-rs checks for this at compile time heh)

File: BunnymodXT/modules/ClientDLL.cpp [hunk 16]
```diff
@@ -66,10 +66,9 @@ extern "C" int __cdecl HUD_UpdateClientData(client_data_t* pcldata, float flTime
 
 extern "C" void __cdecl _ZN20CStudioModelRenderer21StudioCalcAttachmentsEv(void *thisptr)
 {
-	return ClientDLL::HOOKED_StudioCalcAttachments(thisptr);
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.cpp]: This commit sounds like it should be squashed into the corresponding previous commit

File: BunnymodXT/modules/ClientDLL.cpp [hunk 17]
```diff
@@ -119,7 +119,8 @@ void ClientDLL::Hook(const std::wstring& moduleName, void* moduleHandle, void* m
 			ORIG_HUD_DrawTransparentTriangles, HOOKED_HUD_DrawTransparentTriangles,
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.cpp:118]: Also candidate for squashing

File: BunnymodXT/modules/ClientDLL.cpp [hunk 18]
```diff
@@ -119,6 +119,7 @@ void ClientDLL::Hook(const std::wstring& moduleName, void* moduleHandle, void* m
...
 			ORIG_HUD_Key_Event, HOOKED_HUD_Key_Event,
 			ORIG_HUD_UpdateClientData, HOOKED_HUD_UpdateClientData,
+			ORIG_EV_GetDefaultShellInfo, HOOKED_EV_GetDefaultShellInfo,
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.cpp:121]: Another squashing candidate

File: BunnymodXT/cvars.cpp [hunk 19]
```diff
@@ -20,6 +20,8 @@ namespace CVars
 	CVarWrapper bxt_tas_norefresh_until_last_frames("bxt_tas_norefresh_until_last_frames", "0");
 	CVarWrapper bxt_tas_write_log("bxt_tas_write_log", "0");
 	CVarWrapper bxt_tas_playback_speed("bxt_tas_playback_speed", "1");
+	CVarWrapper bxt_disable_vgui("bxt_disable_vgui", "0");
+	CVarWrapper bxt_force_zmax("bxt_force_zmax", "0");
```
Comment (YaLTeR) [inline BunnymodXT/cvars.cpp:24]: Seems unrelated? Is this due to a merge commit? Better to do the rebase thing instead

**Review by YaLTeR — CHANGES_REQUESTED** [review]
tldr: let's try to make good commits

File: BunnymodXT/modules/ClientDLL.cpp (hunk 7 above)
Comment (chinese-soup) [reply BunnymodXT/modules/ClientDLL.cpp:76]: I don't understand review comments like this one, what do you mean, it **is hooked** on the latest commit?

File: BunnymodXT/modules/ClientDLL.cpp (hunk 6 above)
Comment (chinese-soup) [reply BunnymodXT/modules/ClientDLL.cpp]: So wait, you want me to go back to the commits that have mistakes that were fixed by the commit after that and fix the mistakes there or just squash it?

Comment (YaLTeR) [reply BunnymodXT/modules/ClientDLL.cpp]: I want the commits to not contain mistakes in new code. I noted some commits can be squashed into previous completely, if those commits fix mistakes then naturally you'll end up with the original commit without mistakes.

File: BunnymodXT/cvars.cpp (hunk 19 above)
Comment (YaLTeR) [reply BunnymodXT/cvars.cpp:24]: The point is to have good clean commits so I can review a big PR commit-by-commit without being bombarded by commented-out code, mistakes fixed in future commits and unrelated changes from merges.

File: BunnymodXT/modules/ClientDLL.cpp (hunk 6 above)
Comment (chinese-soup) [reply BunnymodXT/modules/ClientDLL.cpp]: alright i get it now

File: BunnymodXT/modules/HwDLL.cpp (hunk 13 above)
Comment (chinese-soup) [reply BunnymodXT/modules/HwDLL.cpp:4706]: ```float fovY = desired_viewmodel_fov;```

Comment (chinese-soup) [reply BunnymodXT/modules/HwDLL.cpp:4706]: yea i shall just float fovY = CVars::bxt_viewmodel_fov.GetFloat();

File: BunnymodXT/modules/ClientDLL.cpp (hunk 3 above)
Comment (chinese-soup) [reply BunnymodXT/modules/ClientDLL.cpp]: From what I understand the player would have to change the FOV on the same frame as zooming the crossbow/357/etc., but I've moved it down just to be sure, yea.

File: BunnymodXT/modules/ClientDLL.cpp [hunk 20]
```diff
@@ -63,6 +63,21 @@ extern "C" int __cdecl HUD_UpdateClientData(client_data_t* pcldata, float flTime
 {
 	return ClientDLL::HOOKED_HUD_UpdateClientData(pcldata, flTime);
 }
+
+extern "C" void __cdecl _ZN20CStudioModelRenderer21StudioCalcAttachmentsEv(void *thisptr)
+{
+	return ClientDLL::HOOKED_StudioCalcAttachments_Linux(thisptr);
+}
+extern "C" void __cdecl _Z15VectorTransformPKfPA4_fPf(float *in1, float *in2, float *out)
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.cpp:71]:
```suggestion

extern "C" void __cdecl _Z15VectorTransformPKfPA4_fPf(float *in1, float *in2, float *out)
```

File: BunnymodXT/modules/ClientDLL.cpp [hunk 21]
```diff
@@ -128,7 +146,9 @@ void ClientDLL::Unhook()
 			ORIG_HUD_Frame,
 			ORIG_HUD_DrawTransparentTriangles,
 			ORIG_HUD_Key_Event,
-			ORIG_HUD_UpdateClientData);
+			ORIG_HUD_UpdateClientData,
+			ORIG_StudioCalcAttachments,
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.cpp:150]:
```suggestion
			ORIG_EV_GetDefaultShellInfo,
			ORIG_StudioCalcAttachments,
```

File: BunnymodXT/stdafx.hpp [hunk 22]
```diff
@@ -58,6 +58,9 @@ using std::ptrdiff_t;
 #include "HLSDK/common/pmtrace.h"
 #include "HLSDK/common/event_api.h"
 #include "HLSDK/common/com_model.h"
+#include "HLSDK/common/cl_entity.h"
+#include "HLSDK/common/event_args.h"
+
```
Comment (YaLTeR) [inline BunnymodXT/stdafx.hpp:63]: Of course no ready for review PR is complete without an extra blank line
```suggestion
```

**Review by YaLTeR — CHANGES_REQUESTED** [review]
Also some warnings on the Windows build:
```
  HwDLL.cpp
D:\a\BunnymodXT\BunnymodXT\BunnymodXT\modules\HwDLL.cpp(4721,39): warning C4244: '=': conversion from 'double' to 'GLfloat', possible loss of data [D:\a\BunnymodXT\BunnymodXT\build\BunnymodXT.vcxproj]
  ClientDLL.cpp
D:\a\BunnymodXT\BunnymodXT\BunnymodXT\modules\ClientDLL.cpp(667,15): warning C4244: 'initializing': conversion from 'double' to 'float', possible loss of data [D:\a\BunnymodXT\BunnymodXT\build\BunnymodXT.vcxproj]
D:\a\BunnymodXT\BunnymodXT\BunnymodXT\modules\ClientDLL.cpp(668,14): warning C4244: 'initializing': conversion from 'double' to 'float', possible loss of data [D:\a\BunnymodXT\BunnymodXT\build\BunnymodXT.vcxproj]
```

File: BunnymodXT/modules/HwDLL.cpp [hunk 23]
```diff
@@ -1330,6 +1403,17 @@ void HwDLL::FindStuff()
...
 		}
 
+		{
+			auto pattern = fR_StudioCalcAttachments.get();
+			if (ORIG_R_StudioCalcAttachments) {
+				EngineDevMsg("[hw dll] Found R_StudioCalcAttachments at %p (using the %s pattern).\n", ORIG_R_StudioCalcAttachments, pattern->name());
+				EngineDevMsg("[hw dll] Found VectorTransform at %p.\n", ORIG_VectorTransform);
+			} else {
+				EngineDevWarning("[hw dll] Could not find R_StudioCalcAttachments.\n");
+				EngineWarning("[client dll] Special effects of weapons will be misplaced when bxt_viewmodel_fov is used.\n");
```
Comment (hobokenn) [inline BunnymodXT/modules/HwDLL.cpp:1413]:
```suggestion
				EngineWarning("[hw dll] Special effects of weapons will be misplaced when bxt_viewmodel_fov is used.\n");
```
my bad

**Comment by chinese-soup** [comment]
> Also some warnings on the Windows build:

Sry didn't check that since it didn't warn in OpenAG and clang-tidy is already always full of warnings so I just don't read that at all. Fixed.

**Review by YaLTeR — APPROVED** [review]

**Comment by YaLTeR** [comment]
Thanks!
